code/integration-tests/runtime-tests/test/tests/pablo/constantProductDex/constantProductTests.ts

```diff
@@ -1,285 +1,1888 @@
+        api.createType("BTreeMap<SafeRpcWrapper, SafeRpcWrapper>", assets)
+      );
+      const {
+        data: [resultWho, resultPoolId, resultAssetsAmount]
+      } = await sendAndWaitForSuccess(
         api,
-        poolId,
-        walletId3,
-        BigInt(transferredTokens.toString()) - Pica(5)
+        walletLpProvider1,
+        api.events.pablo.LiquidityRemoved.is,
+        api.tx.pablo.removeLiquidity(hardCodedPool1.poolId, amountLpToRemove, assets)
+      );
+      // ToDo
+      expect(resultWho.toString()).to.be.equal(api.createType("AccountId32", walletLpProvider1.publicKey).toString());
+      expect(resultPoolId).to.be.bignumber.equal(new BN(hardCodedPool1.poolId.toString()));
+      // @ts-ignore
+      expect(expectedReturnAmount["assets"].toPrimitive()[4]).to.be.equal(resultAssetsAmount.toPrimitive()[4]);
+      // @ts-ignore
+      expect(expectedReturnAmount["assets"].toPrimitive()[130]).to.be.equal(resultAssetsAmount.toPrimitive()[130]);
+      const baseAssetFundsInPoolsAfterRemoval = await api.query.tokens.accounts(
+        hardCodedPool1.poolWalletAddress,
+        hardCodedPool1.baseAssetId
+      );
+      const quoteAssetFundsInPoolsAfterRemoval = await api.query.tokens.accounts(
+        hardCodedPool1.poolWalletAddress,
+        hardCodedPool1.quoteAssetId
+      );
+      // @ts-ignore
+      const returnedAmountBaseAsset = resultAssetsAmount.toPrimitive()[4];
+      // @ts-ignore
+      const returnedAmountQuoteAsset = resultAssetsAmount.toPrimitive()[130];
+      expect(baseAssetFundsInPoolsAfterRemoval.free).to.be.bignumber.equal(
+        baseAssetFundsInPoolsBeforeRemoval.free.sub(new BN(returnedAmountBaseAsset.toString()))
+      );
+      expect(quoteAssetFundsInPoolsAfterRemoval.free).to.be.bignumber.equal(
+        quoteAssetFundsInPoolsBeforeRemoval.free.sub(new BN(returnedAmountQuoteAsset.toString()))
       );
-      // ToDo (D. Roth): Re- enable after pablo rework!
-      // expect(result.resultQuote.toBn()).to.be.bignumber.greaterThan("0");
+      const poolFundsRatioAfterRemoval = BigNumber(baseAssetFundsInPoolsBeforeRemoval.free.toString())
+        .dividedBy(BigNumber(quoteAssetFundsInPoolsBeforeRemoval.free.toString()))
+        .toNumber();
+      expect(poolFundsRatioAfterRemoval).to.be.equal(poolFundsRatioBeforeRemoval);
     });
 
     it(
-      "Given that the users have sufficient balance, " + "users can't buy assets that is not listed in the pool",
+      "#3.7 After adding funds, transferring 50% of them to another wallet " +
+        "if there's been trades in the meantime, both wallets can withdraw their liquidity," +
+        "and will receive equal amounts of funds back.",
+      async function () {
+        this.timeout(5 * 60 * 1000);
+        /*
+          1. Provide liquidity
+          2. Move 50% of liquidity to another wallet.
+          3. Trade
+          4. Remove liquidity of first wallet (expect fees)
+          5. Remove liquidity of second wallet (expect no fees)
```

Comment:
Shouldn't we expect fees for all lp holders, regardless of how they receive the lp tokens.

Reply:
Yes correct, the comment here is wrong. Thanks for pointing out.